Read if needed on NEED_UNWRAP

vkostyukov opened the pull request:
## Motivation:

There are some use cases when a client may only be willing to read from a channel once its previous write is finished (eg: serial dispatchers in [Finagle](https://github.com/twitter/finagle)). In this case, a connection with `SslHandler` installed and `ctx.channel().config().isAutoRead() == false` will stall in 100% of cases no matter what order of "channel active", "write", "flush" events was.

The use case is following (how Finagle serial dispatchers work):

1. Client `writeAndFlush`es and waits on a write-promise to perform `read()` once it's satisfied.
2. A write-promise will only be satisfied once `SslHandler` finishes with handshaking and sends the unencrypted queued message.
3. The handshaking process itself requires a number of `read()`s done by a client but the `SslHandler` doesn't request them explicitly assuming that either auto-read is enabled or client requested at least one `read()` already.
4. At this point a client will stall with `NEED_UNWRAP` status returned from underlying engine.

## Modifications:

Always request a `read()` on `NEED_UNWRAP` returned from engine if a) it's handshaking and b) auto read is disabled and c) it wasn't requested already.

## Result:

We have an internal test that passes with this patch, but stuck establishing a TLS connection without it. I'm happy to add a unit test, but I need some pointers on where to put it.

normanmaurer commented:
@vkostyukov Add it to the test folder of the handler module or in the testsuite module. Also please check why your change produce other testfailures. Thanks!

vkostyukov commented:
Thanks @normanmaurer! I had a look at the failures and I think I have an idea about what's going on. I'l add the test once I fix the existing failures.

normanmaurer commented:
Thanks a lot! Just ping me once ready

> Am 25.03.2016 um 06:46 schrieb Vladimir Kostyukov <notifications@github.com>:
> 
> Thanks @normanmaurer! I had a look at the failures and I think I have an idea about what's going on. I'l add the test once I fix the existing failures.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

normanmaurer commented:
Yay @vkostyukov, tests pass again

vkostyukov commented:
Thanks @normanmaurer! Working on a unit test right now. Will ping you once this PR is updated (should be done before noon in SF).

normanmaurer commented:
Ok cool

normanmaurer commented:
Also once done ensure you update the commit message to be inline wirh our template:

http://netty.io/wiki/writing-a-commit-message.html

vkostyukov commented:
@normanmaurer I added a unit test that tests: channel active, write, flush case. Although I'm having hard times figuring out how to create an `EmbeddedChannel` with auto-read = false so I can test write, flush, channel active events sequence (that's what Finagle does today).

normanmaurer commented:
It's not really possible... Use LocalChannel

> Am 25.03.2016 um 22:25 schrieb Vladimir Kostyukov <notifications@github.com>:
> 
> @normanmaurer I added a unit test that tests: channel active, write, flush case. Although I'm having hard times figuring out how to create an EmbeddedChannel with auto-read = false so I can test write, flush, channel active events sequence (that's what Finagle does today).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

normanmaurer commented:
Or just NioSocketChannel...

vkostyukov commented:
Oh thanks! Let me try that.

vkostyukov commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +65,34 @@ public void testNonByteBufNotPassThrough() throws Exception {
         ch.writeOutbound(new Object());
     }
+
+    private class ReadSnooper extends ChannelOutboundHandlerAdapter {
```
Since we're here. Is there any better way of doing this?

normanmaurer replied:
private static final

normanmaurer replied:
I think this is fine

normanmaurer commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +65,34 @@ public void testNonByteBufNotPassThrough() throws Exception {
+      // SslHandle will start handshaking on handlerAdded
+      ch.pipeline().addLast(new SslHandler(engine));
+
+      ch.writeOutbound(Unpooled.EMPTY_BUFFER); // write and flush
```
Assert the return value

normanmaurer commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +65,34 @@ public void testNonByteBufNotPassThrough() throws Exception {
+
+      assertTrue(snooper.readIssued);
+    }
 }
```
Call channel.finish() and assert the return value

normanmaurer commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +65,34 @@ public void testNonByteBufNotPassThrough() throws Exception {
+
+    @Test
+    public void testIssueReadAfterActiveWriteFlush() throws Exception {
+      SSLEngine engine = SSLContext.getDefault().createSSLEngine();
```
Use SslContextBuilder for all of this

normanmaurer commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +65,34 @@ public void testNonByteBufNotPassThrough() throws Exception {
+
+      ch.pipeline().addFirst(snooper);
+      // SslHandle will start handshaking on handlerAdded
+      ch.pipeline().addLast(new SslHandler(engine));
```
The SslContext build via SslContextBuilder can then be used to create the SslHandler

vkostyukov commented:
@normanmaurer In the meantime, what do you think of adding a new constructor to `EmbeddedChannel` that also accepts `ChannelConfig`?

normanmaurer commented:
I don't think this is needed honestly

> Am 25.03.2016 um 23:20 schrieb Vladimir Kostyukov <notifications@github.com>:
> 
> @normanmaurer In the meantime, what do you think of adding a new constructor to EmbeddedChannel that also accepts ChannelConfig?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

normanmaurer commented:
@vkostyukov seems like there is a leak. I suspect you not release the buffer in your unit test. Run the test with "-Pleak" to see it.

normanmaurer commented:
@vkostyukov you think you will get this done today? I woul like to cut another release tomorrow

vkostyukov commented:
Yeah, I will pick it up in the morning. I guess we're extremely interested in including this into the next release since it blocks our work on TLS in Finagle/Netty 4.

As a side note. I tried using both `LocalChannel` and `NioSocketChannel`, but still not sure I can figure out how to explicitly control the order of events. I want to make sure we test both cases: WritAndFlush and then ChannelActive and vise versa.

@normanmaurer Anyways, thanks a lot for you support. Totally appreciate your time on this.

normanmaurer commented:
Can you explain me exactly what you would like to test ? If I know exactly what you try to do I may be sble to give some tips

> Am 28.03.2016 um 09:01 schrieb Vladimir Kostyukov <notifications@github.com>:
> 
> Yeah, I will pick it up in the morning. I guess we're extremely interested in including this into the next release since it blocks our work on TLS in Finagle/Netty 4.
> 
> As a side note. I tried using both LocalChannel and NioSocketChannel, but still not sure I can figure out how to explicitly control the order of events. I want to make sure we test both cases: WritAndFlush and then ChannelActive and vise versa.
> 
> @normanmaurer Anyways, thanks a lot for you support. Totally appreciate your time on this.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

johnou commented:
Does https://github.com/netty/netty/pull/4254 relate at all?

normanmaurer commented:
Nope

> Am 28.03.2016 um 12:04 schrieb Johno Crawford <notifications@github.com>:
> 
> Does #4254 relate at all?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

vkostyukov commented:
@normanmaurer Thanks! As far as I understand `SslHandler`, there are two different code paths that start a handshake on a client: 1) either `handlerAdded` or `channelActive` 2) `flush` (after `write`). This PR fixes both code paths so the underlying engine is not starving anymore. What I'm trying to test is two different order of events: `flush` -> `channelActive` and `channelActive` -> `flush`. The second order is pretty easy to test. In fact the current state of this PR already does that. I'm having troubles figuring out how to make sure a channel (an `SslHandler` on top of channel) will get `flush` before `channelActive`.

This seems to be pretty simple task to do. And `EmbeddedChannel` (I think) is designed to be able to test handlers in separate.

Okay, wile writing this I figured out something. I will insert a handler right in front of `SslHandler` that drops any `channelActive` events so I can test the first code path. That works in my head, but let me try that in a wild.

Thanks again!

normanmaurer commented:
That will work :)

> Am 28.03.2016 um 18:24 schrieb Vladimir Kostyukov <notifications@github.com>:
> 
> @normanmaurer Thanks! As far as I understand SslHandler, there are two different code paths that start a handshake on a client: 1) either handlerAdded or channelActive 2) flush (after write). This PR fixes both code paths so the underlying engine is not starving anymore. What I'm trying to test is two different order of events: flush -> channelActive and channelActive -> flush. The second order is pretty easy to test. In fact the current state of this PR already does that. I'm having troubles figuring out how to make sure a channel (an SslHandler on top of channel) will get flush before channelActive.
> 
> This seems to be pretty simple task to do. And EmbeddedChannel (I think) is designed to be able to test handlers in separate.
> 
> Okay, wile writing this I figured out something. I will insert a handler right in front of SslHandler that drops any channelActive events so I can test the first code path. That works in my head, but let me try that in a wild.
> 
> Thanks again!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

vkostyukov commented:
@normanmaurer I just updated the PR. Hopefully that leak will go away since I'm not calling `ch.finish()` at the end of the test. Let's see if CI agrees to it.

normanmaurer commented:
Let me check

> Am 28.03.2016 um 19:22 schrieb Vladimir Kostyukov <notifications@github.com>:
> 
> @normanmaurer I just updated the PR. Hopefully that leak will go away since I'm not calling ch.finish() at the end of the test. Let's see if CI agrees to it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

vkostyukov commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +67,52 @@ public void testNonByteBufNotPassThrough() throws Exception {
     }
+
+    private static class TlsReadTest extends ChannelOutboundHandlerAdapter {
+        public volatile boolean readIssued;
```
I guess I can make that private now.

normanmaurer replied:
+1

vkostyukov commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +67,52 @@ public void testNonByteBufNotPassThrough() throws Exception {
+        }
+
+        public void test(final boolean dropChannelActive) throws Exception {
+          SSLEngine engine = SSLContext.getDefault().createSSLEngine();
```
I was thinking about using `SslContextBuilder` here, but decided to stay aligned with the rest of the tests in this file and use `SSLContext.getDefault()`. I hope that's ok, because it doesn't really matter what engine is used in the context of this PR and two new tests.

normanmaurer replied:
ok fair enough

normanmaurer commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +67,52 @@ public void testNonByteBufNotPassThrough() throws Exception {
+
+          assertTrue(ch.writeOutbound(Unpooled.EMPTY_BUFFER));
+          assertTrue(readIssued);
+          assertTrue(ch.finish());
```
You should call `ch.finishAndReleaseAll()` so all buffers are released (and so no leaks)

vkostyukov replied:
Taken. Thanks!

normanmaurer commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +67,52 @@ public void testNonByteBufNotPassThrough() throws Exception {
+                  }
+                }
+              }
+          };
```
Why not just do this directly in the constructor of `EmbeddedChannel` and so not create the array ?

vkostyukov replied:
Good idea! Taken.

normanmaurer commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +67,52 @@ public void testNonByteBufNotPassThrough() throws Exception {
+    @Test
+    public void testIssueReadAfterWriteFlushActive() throws Exception {
+        // the handshake is initiated by flush
+        new TlsReadTest().test(true /*dropChannelActive*/);
```
just remove the comment here.

normanmaurer commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +67,52 @@ public void testNonByteBufNotPassThrough() throws Exception {
+    @Test
+    public void testIssueReadAfterActiveWriteFlush() throws Exception {
+        // the handshake is initiated by channelActive
+        new TlsReadTest().test(false /*dropChannelActive*/);
```
just drop the comment here.

normanmaurer commented on handler/src/main/java/io/netty/handler/ssl/SslHandler.java:
```diff
@@ -562,11 +564,13 @@ private void wrap(ChannelHandlerContext ctx, boolean inUnwrap) throws SSLExcepti
-    private void finishWrap(ChannelHandlerContext ctx, ByteBuf out, ChannelPromise promise, boolean inUnwrap) {
+    private void finishWrap(ChannelHandlerContext ctx, ByteBuf out, ChannelPromise promise, boolean inUnwrap,
+            boolean needUnwrap) {
+
```
remove empty line

normanmaurer commented:
@vkostyukov I think once you addressed my comments the leak is gone and it is ready to be merged.

Scottmitch commented on handler/src/main/java/io/netty/handler/ssl/SslHandler.java:
```diff
@@ -917,16 +927,19 @@ public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
         if (!ctx.channel().config().isAutoRead() && (!firedChannelRead || !handshakePromise.isDone())) {
             // No auto-read used and no message passed through the ChannelPipeline or the handhshake was not complete
             // yet, which means we need to trigger the read to ensure we not encounter any stalls.
             ctx.read();
```
should this just be `read(ctx)` so that `readDuringHandshake` is set consistently?

normanmaurer replied:
actually yes I think this is a good idea.. @vkostyukov please adjust.

vkostyukov replied:
I'd probably need to add `throws Exception` for every single method in the call-graph to `readIfNeeded` including `wrap`, `finishWrap`, `unwrap`. If you're ok with that - I can do it. Otherwise, we can handle `readDuringHandshake = true;` explicitly inside `readIfNeeded`.

vkostyukov replied:
One more thing. I'm not 100% sure we need to change `readDuringHandshake` via `readIfNeeded` since it serves a different propose. I'm don't think we should count the read `SslHandler` initiated to handle the handshake as `readDuringHandshake`. Otherwise it should be `true` all the the time. Although, I'm pretty sure I'm missing something here.

normanmaurer replied:
@vkostyukov ol just leave it as it is for now. We can review this as a new issue as the code was like this before. @Scottmitch ok with you? I will investigate if this is ok

vkostyukov replied:
Right. I think that's what we've been doing before. My take it that a read initiated by `SslHanlder` to finish the handshake shouldn't affect the `readDuringHandshake` flag. Although, I can see some sense from the opposite relation. Maybe we don't need to trigger a `read()` if `readDuringHandshake` is true. But again, we don't check that today.

Scottmitch replied:
find to leave as is for now. we can dig deeper and handle as a followup PR.

Scottmitch replied:
> I'd probably need to add throws Exception for every single method in the call-graph

@vkostyukov - You should be able to remove `throws Exception` from the signature of `public void read(ChannelHandlerContext ctx)` ... however to your point it seems reasonable that a read "synthesized" by SslHandler should not force a read after the handshake is complete.

normanmaurer replied:
We can not remove throws Exception here as this would break a users handler that extend SslHandler and override read(...) and still have the throws in the signature. But anyway let us leave it as it is for now and investigate later

> Am 28.03.2016 um 22:40 schrieb Scott Mitchell <notifications@github.com>:
> 
> In handler/src/main/java/io/netty/handler/ssl/SslHandler.java:
> 
> >  
> > +    private void readIfNeeded(ChannelHandlerContext ctx) {
> >          // If handshake is not finished yet, we need more data.
> >          if (!ctx.channel().config().isAutoRead() && (!firedChannelRead || !handshakePromise.isDone())) {
> >              // No auto-read used and no message passed through the ChannelPipeline or the handhshake was not complete
> >              // yet, which means we need to trigger the read to ensure we not encounter any stalls.
> >              ctx.read();
> I'd probably need to add throws Exception for every single method in the call-graph
> 
> @vkostyukov - You should be able to remove throws Exception from the signature of public void read(ChannelHandlerContext ctx) ... however to your point it seems reasonable that a read "synthesized" by SslHandler should not force a read after the handshake is complete.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

Scottmitch replied:
> this would break a users handler that extend SslHandler and override read(...)

Good point.

Scottmitch commented on handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java:
```diff
@@ -63,4 +67,52 @@ public void testNonByteBufNotPassThrough() throws Exception {
         ch.writeOutbound(new Object());
     }
+
+    private static class TlsReadTest extends ChannelOutboundHandlerAdapter {
```
nit: `final`

vkostyukov replied:
Done!

Scottmitch commented:
@vkostyukov - thanks for the PR! after comments address lgtm.

vkostyukov commented:
@normanmaurer, @Scottmitch thank you for the review!

I've updated the PR and addressed all the comments except for the `readDuringHandshake` one. I'm happy to spend some time to try to figure out what do with that (maybe check that flag in `readIfNeeded` in some way), if you think this is something we need to address as part of this PR.

mosesn commented:
LGTM

normanmaurer commented:
Cherry-picked into 4.1 (84bbbf7e09513e761496fe1faf3407fc06a32228) and 4.0 (ea4a8e339c54f1e4eb3a047f397d0801d53ff0a6).

@vkostyukov thanks a lot!
